[CSS] @font-palette-values family-name supports multiple values

mdubet opened the pull request:
#### 83db139305cad9bc26aae1844c44a43b12e5d986
<pre>
[CSS] @font-palette-values family-name supports multiple values
<a href="https://bugs.webkit.org/show_bug.cgi?id=252569">https://bugs.webkit.org/show_bug.cgi?id=252569</a>
rdar://105975619

Reviewed by Tim Nguyen.

This patch allows @font-palette-values family-name descriptor to have
a list of family names (per spec).

<a href="https://drafts.csswg.org/css-fonts-4/#family-name-syntax">https://drafts.csswg.org/css-fonts-4/#family-name-syntax</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-36-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-36-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-36.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-palette-values-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-palette-values-valid-expected.txt:
* Source/WebCore/css/CSSFontPaletteValuesRule.cpp:
(WebCore::CSSFontPaletteValuesRule::fontFamily const):
(WebCore::CSSFontPaletteValuesRule::cssText const):
(WebCore::CSSFontPaletteValuesRule::~CSSFontPaletteValuesRule): Deleted.
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontPaletteValuesRule):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleFontPaletteValues::create):
(WebCore::StyleRuleFontPaletteValues::StyleRuleFontPaletteValues):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
(WebCore::CSSParserImpl::consumeFontPaletteValuesRule):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameListRaw):
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameList):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:

Canonical link: <a href="https://commits.webkit.org/267411@main">https://commits.webkit.org/267411@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/906090869eecf2c6b17ca36162bfe50eeb9862c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16481 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18258 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15455 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16946 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17807 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16678 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17094 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14250 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19034 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14338 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21728 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14218 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15326 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15099 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19412 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15718 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15695 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13326 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18038 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14896 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3953 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19265 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19258 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15521 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4071 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4874,10 +4874,13 @@ AtomString consumeFamilyNameRaw(CSSParserTokenRange& range)
     if (range.peek().type() != IdentToken)
         return nullAtom();
+    // <family-name> can't contain unquoted generic families
+    if (!genericFontFamily(range.peek().id()).isNull())
```
Would be good to add a parsing subtest for this.

mdubet replied:
There is already a parsing test for that.

nt1m replied:
That we reject generic font families?

mdubet replied:
Yes https://github.com/web-platform-tests/wpt/blob/29bc632bd38bf33bd475ef5b925fcea029cf6249/css/css-fonts/parsing/font-palette-values-invalid.html#L22

nt1m replied:
ah sounds good, thanks!

nt1m reviewed (commented):
Can we add a reftest for this feature?

nt1m reviewed (approved):
Please export this WPT

nt1m reviewed (changes requested):
Some failures need to be looked at

mdubet commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-palette-values-invalid-expected.txt:
```diff
@@ -1,6 +1,6 @@
 
 PASS CSS Fonts Module Level 4: parsing @font-palette-values
-PASS CSS Fonts Module Level 4: parsing @font-palette-values 1
```
We have another issue here (which was hidden by our bug of bug) https://bugs.webkit.org/show_bug.cgi?id=260066

mdubet committed:
[CSS] @font-palette-values family-name supports multiple values
https://bugs.webkit.org/show_bug.cgi?id=252569
rdar://105975619

Reviewed by Tim Nguyen.

This patch allows @font-palette-values family-name descriptor to have
a list of family names (per spec).

https://drafts.csswg.org/css-fonts-4/#family-name-syntax

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-36-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-36-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-36.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-palette-values-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-palette-values-valid-expected.txt:
* Source/WebCore/css/CSSFontPaletteValuesRule.cpp:
(WebCore::CSSFontPaletteValuesRule::fontFamily const):
(WebCore::CSSFontPaletteValuesRule::cssText const):
(WebCore::CSSFontPaletteValuesRule::~CSSFontPaletteValuesRule): Deleted.
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontPaletteValuesRule):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleFontPaletteValues::create):
(WebCore::StyleRuleFontPaletteValues::StyleRuleFontPaletteValues):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
(WebCore::CSSParserImpl::consumeFontPaletteValuesRule):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameListRaw):
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameList):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:

Canonical link: https://commits.webkit.org/267411@main
